File: app/src/main/java/com/nacare/capture/data/adapters/TreeAdapter.java
package com.nacare.capture.data.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.nacare.capture.R;
import com.nacare.capture.data.model.OrgTreeNode;

import java.util.List;

public class TreeAdapter extends RecyclerView.Adapter<TreeAdapter.TreeViewHolder> {

    private final Context context;
    private final List<OrgTreeNode> treeNodes;
    private final OnItemClickListener click;
    private boolean isChecked;

    public TreeAdapter(Context context, List<OrgTreeNode> treeNodes, OnItemClickListener click) {
        this.context = context;
        this.treeNodes = treeNodes;
        this.click = click;
    }


    public interface OnItemClickListener {
        void onItemClick(OrgTreeNode orgTreeNode);
    }

    public class TreeViewHolder extends RecyclerView.ViewHolder {
        public TreeViewHolder(View itemView) {
            super(itemView);
        }
    }

    @Override
    public TreeViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.tree_item_layout, parent, false);
        return new TreeViewHolder(view);
    }

    @Override
    public void onBindViewHolder(TreeViewHolder holder, int position) {
        OrgTreeNode node = treeNodes.get(position);
        View itemView = holder.itemView;

        // Adjust the indentation based on the tree level
        int paddingLeft = 20 * getTreeLevel(node);
        itemView.setPadding(paddingLeft, 0, 0, 0);

        // Set label text
        TextView labelTextView = itemView.findViewById(R.id.org_unit_name);
        labelTextView.setText(node.getLabel());

        // Handle click to expand/collapse
        itemView.setOnClickListener(v -> toggleNode(position, holder));

        // Handle visibility based on whether the node has children and is expanded
        ImageView arrowIcon = itemView.findViewById(R.id.org_unit_icon);
        arrowIcon.setVisibility(node.getChildren().isEmpty() ? View.INVISIBLE :
                (node.getLevel().equals("5") ? View.INVISIBLE : View.VISIBLE));

        // Set label text
        CheckBox checkbox = itemView.findViewById(R.id.checkbox);
        checkbox.setVisibility(node.getChildren().isEmpty() ? (node.getLevel().equals("5") ? View.VISIBLE : View.INVISIBLE) : View.INVISIBLE);

        // Handle visibility based on whether the node has children and is expanded
        if (node.getChildren().isEmpty()) {
            checkbox.setOnClickListener(v -> {
                OrgTreeNode org = treeNodes.get(position);
                if (org.getLevel().equals("5")) {
                    click.onItemClick(org);
                }
            });
        }

        // Handle expand/collapse state
        arrowIcon.setImageResource(node.isExpanded() ? R.drawable.baseline_remove_circle_24 : R.drawable.ic_add_circle);

        // Set up child RecyclerView
        RecyclerView childRecyclerView = itemView.findViewById(R.id.childRecyclerView);
        TreeAdapter childAdapter = new TreeAdapter(context, node.getChildren(), click);
        childRecyclerView.setAdapter(childAdapter);
        childRecyclerView.setLayoutManager(new LinearLayoutManager(context));

        // Set child RecyclerView visibility based on the expand/collapse state
        childRecyclerView.setVisibility(node.isExpanded() ? View.VISIBLE : View.GONE);
    }

    @Override
    public int getItemCount() {
        return treeNodes.size();
    }

    private int getTreeLevel(OrgTreeNode node) {
        int level = 0;
        OrgTreeNode currentNode = node;
        while (!currentNode.getChildren().isEmpty()) {
            level++;
            currentNode = currentNode.getChildren().get(0);
        }
        return level;
    }

    private void toggleNode(int position, TreeViewHolder holder) {
        OrgTreeNode node = treeNodes.get(position);
        node.setExpanded(!node.isExpanded());
        notifyItemChanged(position);

        // Update child RecyclerView visibility
        RecyclerView childRecyclerView = holder.itemView.findViewById(R.id.childRecyclerView);
        childRecyclerView.setVisibility(node.isExpanded() ? View.VISIBLE : View.GONE);
    }
}

File: app/src/main/java/com/intellisoft/nacare/adapter/TreeAdapter.kt
package com.intellisoft.nacare.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.intellisoft.nacare.helper_class.OrgTreeNode
import com.nacare.ke.capture.R

class TreeAdapter(
    private val context: Context,
    private val treeNodes: List<OrgTreeNode>,
    private val click: (OrgTreeNode) -> Unit
) :
    RecyclerView.Adapter<TreeAdapter.TreeViewHolder>() {

    inner class TreeViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TreeViewHolder {
        val view = LayoutInflater.from(context)
            .inflate(R.layout.tree_item_layout, parent, false)
        return TreeViewHolder(view)
    }

    override fun getItemCount(): Int {
        return treeNodes.size
    }

    override fun onBindViewHolder(holder: TreeViewHolder, position: Int) {
        val node = treeNodes[position]
        val itemView = holder.itemView

        // Adjust the indentation based on the tree level
        val paddingLeft = 20 * getTreeLevel(node)
        itemView.setPadding(paddingLeft, 0, 0, 0)

        // Set label text
        val labelTextView: TextView = itemView.findViewById(R.id.labelTextView)
        labelTextView.text = node.label

        // Handle click to expand/collapse
        itemView.setOnClickListener {
            toggleNode(position, holder)
        }

        // Handle visibility based on whether the node has children and is expanded
        val arrowIcon: ImageView = itemView.findViewById(R.id.arrowIcon)
        arrowIcon.visibility = if (node.children.isNotEmpty()) View.VISIBLE else View.INVISIBLE
        // Handle visibility based on whether the node has children and is expanded
        if (node.children.isEmpty()) {
            labelTextView.setOnClickListener {
                val patient = treeNodes[position]
                click(patient)
            }
        }
        // Handle expand/collapse state
        arrowIcon.rotation = if (node.isExpanded) 90f else 0f

        // Set up child RecyclerView
        val childRecyclerView: RecyclerView = itemView.findViewById(R.id.childRecyclerView)
        val childAdapter = TreeAdapter(context, node.children, click)
        childRecyclerView.adapter = childAdapter
        childRecyclerView.layoutManager = LinearLayoutManager(context)

        // Set child RecyclerView visibility based on the expand/collapse state
        childRecyclerView.visibility = if (node.isExpanded) View.VISIBLE else View.GONE
    }

    private fun getTreeLevel(node: OrgTreeNode): Int {
        var level = 0
        var currentNode = node
        while (currentNode.children.isNotEmpty()) {
            level++
            currentNode = currentNode.children[0]
        }
        return level
    }

    private fun toggleNode(position: Int, holder: TreeViewHolder) {
        val node = treeNodes[position]
        node.isExpanded = !node.isExpanded
        notifyItemChanged(position)

        // Update child RecyclerView visibility
        val childRecyclerView: RecyclerView = holder.itemView.findViewById(R.id.childRecyclerView)
        childRecyclerView.visibility = if (node.isExpanded) View.VISIBLE else View.GONE
    }
}
